### PR TITLE
[HW] Added simple canonicalizers to `hw.struct_inject` and `hw.struct_extract`

### PR DESCRIPTION
**Pull request by nandor**
Canonicalized two injects to the same field to a single inject of the second value.
Canonicalized an extract of an inject to the injected value.

### CODE REVIEW

File: lib/Dialect/HW/HWOps.cpp [hunk 1]
```diff
@@ -1899,17 +1899,25 @@ void StructExtractOp::build(OpBuilder &builder, OperationState &odsState,
...
 
-// A struct extract of a struct create -> corresponding struct create operand.
 OpFoldResult StructExtractOp::fold(ArrayRef<Attribute> operands) {
-  auto structCreate =
-      dyn_cast_or_null<StructCreateOp>(getInput().getDefiningOp());
-  if (!structCreate)
+  auto inputOp = getInput().getDefiningOp();
+  // A struct extract of a struct create -> corresponding struct create operand.
+  if (auto structCreate = dyn_cast_or_null<StructCreateOp>(inputOp)) {
+    if (!structCreate)
```
Comment (darthscsi) [inline lib/Dialect/HW/HWOps.cpp:1906]: redundant check

File: lib/Dialect/HW/HWOps.cpp [hunk 2]
```diff
@@ -1899,17 +1899,25 @@ void StructExtractOp::build(OpBuilder &builder, OperationState &odsState,
...
+      return {};
+    auto ty = type_cast<StructType>(getInput().getType());
+    if (!ty)
```
Comment (darthscsi) [inline lib/Dialect/HW/HWOps.cpp:1909]: redundant check

File: lib/Dialect/HW/HWOps.cpp [hunk 3]
```diff
@@ -1956,6 +1964,23 @@ void StructInjectOp::print(OpAsmPrinter &printer) {
   printer << " : " << getInput().getType();
 }
 
+LogicalResult StructInjectOp::canonicalize(StructInjectOp op,
+                                           PatternRewriter &rewriter) {
+  auto inputOp = op.getInput().getDefiningOp();
+
+  // b = inject(inject(x["field"], v0)["field"], v1) => inject(x["field"], v1)
+  if (auto structInject = dyn_cast_or_null<StructInjectOp>(inputOp)) {
```
Comment (darthscsi) [inline lib/Dialect/HW/HWOps.cpp:1972]: In a graph region, couldn't this cause an infinite loop?
`%a = inject(%a["field], v)`

Comment (nandor) [reply lib/Dialect/HW/HWOps.cpp:1972]: `%a = inject((inject(%a["field], v))["field], v) would canonicalize to `%a = inject(%a["field], v)`, so there is a struct reduction step. If we only see `%a = inject(%a["field], v)` then nothing happens.

Comment (uenoku) [reply lib/Dialect/HW/HWOps.cpp:1984]: I agree with Lenharth, this IR causes an infinite loop.
```
hw.module @ExtractOfInject(%a: !hw.struct<a: i1>, %v: i1) -> (result: i1) {
  %b = hw.struct_inject %b["a"], %v : !hw.struct<a: i1>
  %c = hw.struct_extract %b["a"] : !hw.struct<a: i1>
  // CHECK: hw.output %v
  hw.output %c : i1
}
```

Comment (nandor) [reply lib/Dialect/HW/HWOps.cpp:1984]: I stand corrected. Fixed the cycle.

**Review by uenoku — APPROVED** [review]
LGTM

**Pull request merged**